### PR TITLE
[memory][core] Fix erasing sc-connectors during iterating

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 1]
```diff
@@ -299,6 +299,66 @@ TEST_F(ScMemoryTest, CreateEdges)
...
+    ++i;
+  }
+
+  EXPECT_EQ(i, edgesCount);
+}
+
+TEST_F(ScMemoryTest, IterateEraseEdgesFromNode)
+{
+  ScAddr const classAddr = m_ctx->CreateNode(ScType::NodeConstClass);
+  ScAddr const nodeAddr = m_ctx->CreateNode(ScType::NodeConst);
```
Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:325]: can you create test with some edges in node1Addr and some edges in node2Addr?

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 2]
```diff
@@ -299,6 +299,66 @@ TEST_F(ScMemoryTest, CreateEdges)
...
 }
 
+TEST_F(ScMemoryTest, IterateEraseEdgesBetweenNodes)
+{
+  ScAddr const classAddr = m_ctx->CreateNode(ScType::NodeConstClass);
+  ScAddr const nodeAddr = m_ctx->CreateNode(ScType::NodeConst);
+  
+  size_t const edgesCount = 5;
+  for (size_t i = 0; i < edgesCount; ++i)
+    m_ctx->CreateEdge(ScType::EdgeAccessConstPosPerm, classAddr, nodeAddr);
```
Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:309]:
```suggestion
    m_ctx->CreateEdge(ScType::EdgeAccessConstPosTemp?, classAddr, nodeAddr);
```

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 3]
```diff
@@ -299,6 +299,66 @@ TEST_F(ScMemoryTest, CreateEdges)
...
+
+  size_t i = 0;
+  ScIterator3Ptr const it = m_ctx->Iterator3(classAddr, ScType::EdgeAccessConstPosPerm, ScType::NodeConst);
+  while (it->Next())
+  {
+    m_ctx->EraseElement(it->Get(1));
+    ++i;
+  }
+
+  EXPECT_EQ(i, edgesCount);
```
Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:339]:
```suggestion
  EXPECT_EQ(i, edgesCount);
  EXPECT_EQ(context.getOutputEdgesCount(classAddr), 0u);
```

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 4]
```diff
@@ -299,6 +299,66 @@ TEST_F(ScMemoryTest, CreateEdges)
...
+  ScIterator3Ptr const it = m_ctx->Iterator3(ScType::NodeConstClass, ScType::EdgeAccessConstPosPerm, nodeAddr);
+  while (it->Next())
+  {
+    m_ctx->EraseElement(it->Get(1));
+    ++i;
+  }
+
+  EXPECT_EQ(i, edgesCount);
+}
+
```
Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:361]: what about deleting it->Get(2) while iterating?

Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:361]: also can you check what will happen if 2 edges are removed using first iterator, and then rest of edges are removed using second iterator

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp (hunk 2 above)
Comment (NikitaZotov) [reply sc-memory/tests/sc-memory/common/test_sc_elements.cpp:309]: Edge types don't influence searching and erasing sc-elements

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 5]
```diff
@@ -299,6 +299,156 @@ TEST_F(ScMemoryTest, CreateEdges)
...
+  size_t i = 0;
+  ScIterator3Ptr const it = m_ctx->Iterator3(classAddr, ScType::EdgeAccessConstPosPerm, ScType::NodeConst);
+  while (it->Next())
+  {
+    m_ctx->EraseElement(it->Get(2));
+    ++i;
+  }
+
+  EXPECT_EQ(i, 2u);
```
Comment (kilativ-dotcom) [inline sc-memory/tests/sc-memory/common/test_sc_elements.cpp:448]:
```suggestion
  EXPECT_EQ(i, 2u);
  EXPECT_FALSE(context.IsElement(node1 and node2));
```

**Review by kilativ-dotcom — APPROVED** [review]
works ok :grinning: :partying_face:

**Review by MksmOrlov — APPROVED** [review]

**Pull request merged**